### PR TITLE
feat(exports): add chunked job pipeline for large audience exports

### PR DESCRIPTION
**Pull request by FusionBrah**
Issue: #2092

This is a revised implementation addressing feedback from #2533.

# Description

## Problem

Large audience exports (followers/subscribers CSV) timeout because the entire export happens in a single Sidekiq job. Users with large audiences never receive their export emails.

## Solution

Implement a chunked export pipeline (following the existing `SalesExport` pattern):

- **CreateAndEnqueueChunksJob** - splits audience into 1000-member chunks using `in_batches` (memory-efficient)
- **ProcessChunkJob** - processes each chunk independently with `lock: :until_executed`
- **CompileChunksJob** - assembles final CSV and sends email

Small audiences (≤1000 members) continue using the existing synchronous path for efficiency.

**Key implementation details:**
- Uses `JsonData` module with `attr_json_data_accessor` for options (per maintainer feedback)
- Uses `in_batches` for memory-efficient chunking (no `pluck` into memory)
- Uses `scopes.reduce(&:or)` pattern for query building
- All jobs use `lock: :until_executed` for idempotency
- Follows Rails conventions for association naming (`audience_export_id`)

---

# Before/After

Backend-only change

---

# Test Results

Screenshot 2026-01-23 at 3 02 46 pm

---

# Checklist

- [x] I have read the contributing guidelines
- [x] I have watched Gumroad PR review livestreams
- [x] I have performed a self-review and left review comments on my PR
- [x] I have added/updated tests for my changes

---

# AI Disclosure

AI (Claude Code) was used for:
- Code implementation (Sidekiq job classes, chunking logic, CSV compilation)
- Addressing maintainer feedback from PR #2533 (JsonData module, in_batches, scopes pattern)
- Test writing (RSpec specs including edge cases)
- Code review and debugging

### CODE REVIEW

File: app/models/audience_export.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class AudienceExport < ApplicationRecord
+  include JsonData
```
Comment (FusionBrah) [inline app/models/audience_export.rb:4]: Using `JsonData` module with `attr_json_data_accessor` per maintainer feedback on PR #2533. This replaces the old YAML serialization pattern and is the current codebase convention.

File: app/models/audience_export.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  belongs_to :seller, class_name: "User"
+  belongs_to :recipient, class_name: "User"
+  has_many :audience_export_chunks, dependent: :delete_all
+
+  attr_json_data_accessor :followers, default: false
+  attr_json_data_accessor :customers, default: false
+  attr_json_data_accessor :affiliates, default: false
+
+  validate :at_least_one_audience_type_selected
```
Comment (FusionBrah) [inline app/models/audience_export.rb:14]: Custom validation ensures at least one audience type is selected. This matches the validation in `AudienceExportService` and prevents empty exports.

File: app/models/audience_export.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class AudienceExport < ApplicationRecord
+  include JsonData
+
+  belongs_to :seller, class_name: "User"
+  belongs_to :recipient, class_name: "User"
+  has_many :audience_export_chunks, dependent: :delete_all
```
Comment (FusionBrah) [inline app/models/audience_export.rb:8]: Using `dependent: :delete_all` instead of `:destroy` for performance - chunks contain large serialized data and don't have callbacks. Follows the `SalesExport` pattern.

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class Exports::Audience::CreateAndEnqueueChunksJob
+  include Sidekiq::Job
+  sidekiq_options retry: 5, queue: :low, lock: :until_executed
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:5]: `lock: :until_executed` ensures idempotency per CLAUDE.md guidelines. The job also clears stale chunks on retry (line 17) as a secondary safety measure.

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  private
+    def create_chunks
+      @export.audience_export_chunks.in_batches(of: 100).delete_all
+
+      if build_query.none?
+        Exports::Audience::CompileChunksJob.perform_async(@export.id)
+        return
+      end
+
+      build_query.select(:id).order(:id).in_batches(of: CHUNK_SIZE) do |members|
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:24]: Using `in_batches` per maintainer feedback - avoids loading all member IDs into memory with `pluck(:id)`. Each batch of 1000 IDs is processed then discarded, keeping memory usage constant regardless of audience size.

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    def build_query
+      query = @export.seller.audience_members
+
+      scopes = []
+      scopes << query.where(follower: true) if @export.followers
+      scopes << query.where(customer: true) if @export.customers
+      scopes << query.where(affiliate: true) if @export.affiliates
+
+      scopes.any? ? scopes.reduce(&:or) : query.none
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:42]: Using `scopes.reduce(&:or)` pattern per maintainer feedback. Returns `query.none` when no options selected (defensive, though model validation prevents this state).

File: app/sidekiq/exports/audience/process_chunk_job.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      members = AudienceMember.where(id: @chunk.member_ids)
+                              .select(:email, :min_created_at)
+                              .order(:min_created_at)
+
+      members_data = members.map { |m| [m.email, m.min_created_at&.to_s] }
+
+      @chunk.update!(members_data: members_data, processed: true)
+    end
+
+    def trigger_compile_if_complete
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/process_chunk_job.rb:26]: Only the last chunk to finish triggers CompileChunksJob. The `lock: :until_executed` on CompileChunksJob handles the rare race condition where two chunks finish simultaneously.

File: app/sidekiq/exports/audience/compile_chunks_job.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  private
+    def generate_csv
+      timestamp = Time.current.strftime("%Y-%m-%d-%H-%M-%S")
+      @filename = "Subscribers-#{@export.seller.username}_#{timestamp}.csv"
+
+      tempfile = Tempfile.new(["Subscribers", ".csv"], encoding: "UTF-8")
+
+      CSV.open(tempfile, "wb", headers: FIELDS, write_headers: true) do |csv|
+        @export.audience_export_chunks.order(:id).find_each(batch_size: 1) do |chunk|
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/compile_chunks_job.rb:27]: `batch_size: 1` is intentional - each chunk contains large serialized YAML data. Loading one chunk at a time keeps memory usage low. This follows the `SalesExport` pattern.

File: app/sidekiq/exports/audience/compile_chunks_job.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    @export = AudienceExport.find(export_id)
+
+    tempfile = generate_csv
+    send_email(tempfile)
+    cleanup
+  end
+
+  private
+    def generate_csv
+      timestamp = Time.current.strftime("%Y-%m-%d-%H-%M-%S")
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/compile_chunks_job.rb:21]: Using `strftime` directly per maintainer feedback - cleaner than `to_fs(:db).gsub(...)` pattern.
